### PR TITLE
Database support

### PR DESCRIPTION
**Pull request by ubuntolog**
The GEF is now using an SQLite database to store information about jobs and services. The new package called "db" handles all interactions with the database through the library called GORP. If necessary, any other SQL database can be used

### CODE REVIEW

File: backend-docker/pier/db/db.go [hunk 1]
```diff
@@ -0,0 +1,551 @@
+package db
+
+import (
+	"bytes"
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: please use goimports for nice formatting: https://godoc.org/golang.org/x/tools/cmd/goimports

File: backend-docker/pier/db/db.go [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+	Name      string
+	Path      string
+	IsInput   bool
+	Revision  int
+	ServiceID string
+}
+
+// InitDb initializes the database engine
+func InitDb() (Db, error) {
+	var dataBaseHandler Db
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: it's more difficult to trace the usage of this variable, and it's not really needed (use return Db{nil}, err or return Db{*dataBaseMap}, err)

File: backend-docker/pier/db/db.go [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+	err := d.SelectOne(&serviceFromTable, "SELECT * FROM services WHERE ID=?", string(serviceID))
+
+	if err == nil {
+		service, err = d.MapStoredService2JSON(ServiceID(serviceFromTable.ID), serviceFromTable)
+	}
+
+	return service, err
+}
+
+// NewServiceFromImage extracts metadata and creates a valid GEF service
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: this function and the next one should not be in this file, they have no relation to the db

File: backend-docker/pier/db/db.go [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+// AddJob adds a job to the database
+func (d *Db) AddJob(job Job) error {
+	storedJob := d.MapJSON2StoredJob(job)
+	return d.Insert(&storedJob)
+}
+
+// RemoveJob removes a job and all corresponding tasks from the database
+func (d *Db) RemoveJob(jobID JobID) error {
+	_, err := d.Exec("DELETE FROM Tasks WHERE jobID=?", string(jobID))
+	if err == nil {
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: it's more common to test the error and return if it's not null, so the error cases get out of the way:
`if err != nil { return err }`

File: backend-docker/pier/db/db.go [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+	return err
+}
+
+// RemoveJobTask removes a task from the database
+func (d *Db) RemoveJobTask(taskID string) error {
+	_, err := d.Exec("DELETE FROM Tasks WHERE ID=?", taskID)
+	return err
+}
+
+// MapStoredJob2JSON performs mapping of the database job table to its JSON representation
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: The function name is misleading. `Job` is not JSON, it's the public datastructure that we use. I'd name this function just `JobTable2Job`.

This function should not be public.

File: backend-docker/pier/db/db.go [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+func (d *Db) MapStoredJob2JSON(jobID JobID, storedJob JobTable) (Job, error) {
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: `JobTable` already has a `jobID`, do we need 2 parameters here?

File: backend-docker/pier/db/db.go [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+	job.Input = storedJob.Input
+	job.Created = storedJob.Created
+	job.State = &jobState
+	job.InputVolume = VolumeID(storedJob.InputVolume)
+	job.OutputVolume = VolumeID(storedJob.OutputVolume)
+	job.Tasks = linkedTasks
+	return job, err
+}
+
+// MapJSON2StoredJob performs mapping of the job JSON representation to its database representation
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: naming: Job2JobTable
does not need to be public

File: backend-docker/pier/db/db.go [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+type JobState struct {
+	Error  string
+	Status string
+	Code   int
+}
+
+// JobID exported
+type JobID string
+
+// JobTable stores the information about a service execution (used to store data in a database)
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: This datatype (as all the other ones related to db) does not need to be public, it should be internal to this package (lower case).

File: backend-docker/pier/db/db.go [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+	newTask.Name = taskName
+	newTask.ContainerID = string(taskContainer)
+	newTask.Error = taskError
+	newTask.ExitCode = taskExitCode
+	newTask.ConsoleOutput = taskConsoleOutput.String()
+	newTask.JobID = string(jobID)
+	return d.Insert(&newTask)
+}
+
+// MapStoredService2JSON performs mapping of the database service table to its JSON representation
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: naming, visibility

File: backend-docker/pier/db/db.go [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+	service.Created = storedService.Created
+	service.Size = storedService.Size
+	service.Input = inputPorts
+	service.Input = inputPorts
+	service.Output = outputPorts
+
+	return service, err
+}
+
+// MapJSON2StoredService performs mapping of the service JSON representation to its database representation
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: naming, visibility

File: backend-docker/pier/pier.go [hunk 11]
```diff
@@ -3,178 +3,177 @@ package pier
...
+
 	if err != nil {
 		return nil, def.Err(err, "Cannot create docker client")
 	}
 
 	pier := Pier{
 		docker:   docker,
-		services: NewServiceList(),
-		jobs:     NewJobList(),
+		dataBase: dataBase,
```
Comment (emanueldima) [inline backend-docker/pier/pier.go]: this name is not good, either `db` or `database`

File: backend-docker/pier/db/db.go [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+const GorpVersionColumn = "Revision"           // Column in a table used to keep an internal version number of GORP
+const SQLiteDataBasePath = "gef_db.bin"
+
+// VolumeID contains a docker volume ID
+type VolumeID dckr.VolumeID
+
+// Db is used to keep DbMap
+type Db struct{ gorp.DbMap }
+
+// Job stores the information about a service execution (used to serialize JSON)
```
Comment (emanueldima) [inline backend-docker/pier/db/db.go]: I would say that Job, Service and the main datatypes should be in `pier`, not embedded here. In this way they are part of the official API.

File: backend-docker/pier/db/db.go (hunk 8 above)
Comment (ubuntolog) [inline backend-docker/pier/db/db.go]: We use JobID and ServiceID in pier.go and in api.go
